File: solutions/_75_sort_colors/solution.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function (nums) {
  let lo = 0;
  let mid = 0;
  let hi = nums.length - 1;

  while (mid <= hi) {
    switch (nums[mid]) {
      case 0:
        [nums[lo], nums[mid]] = [nums[mid], nums[lo]];
        lo++;
        mid++;
        break;
      case 1:
        mid++;
        break;
      case 2:
        [nums[mid], nums[hi]] = [nums[hi], nums[mid]];
        hi--;
        break;

      default:
        throw new Error("Invalid color");
    }
  }
};

module.exports = { sortColors };

/*
https://leetcode.com/problems/sort-colors/
*/

File: solutions/_75_sort_colors/solution.java
package _75_sort_colors;

class Solution {

  public void sortColors(int[] nums) {
    int lo = 0;
    int mid = 0;
    int hi = nums.length - 1;
    while (mid <= hi) {
      switch (nums[mid]) {
        case 0:
          swap(nums, mid, lo);
          lo++;
          mid++;
          break;
        case 1:
          mid++;
          break;
        case 2:
          swap(nums, mid, hi);
          hi--;
          break;
        default:
          throw new IllegalArgumentException("Unexpected value: " + nums[mid]);
      }
    }
  }

  public void swap(int[] nums, int i, int j) {
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
  }
}
/*
https://leetcode.com/problems/sort-colors/
*/
